api/fsi.go

```diff
@@ -293,27 +265,53 @@ func (h *FSIHandlers) RestoreHandler(routePrefix string) http.HandlerFunc {
 
 func (h *FSIHandlers) restoreHandler(routePrefix string) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		ref, err := lib.DsRefFromPath(r.URL.Path[len(routePrefix):])
+		method := "fsi.restore"
+		p := h.inst.NewInputParam(method)
+
+		// TODO(dustmop): Add this to UnmarshalParams for methods that can
+		// receive a refstr in the URL, or annotate the param struct with
+		// a tag and marshal the url to that field
+		err := addDsRefFromURL(r, routePrefix)
 		if err != nil {
-			util.WriteErrResponse(w, http.StatusBadRequest, fmt.Errorf("bad reference: %s", err.Error()))
+			util.WriteErrResponse(w, http.StatusBadRequest, err)
 			return
 		}
 
-		// Add the path for the version to restore
-		ref.Path = r.FormValue("path")
-
-		p := &lib.RestoreParams{
-			Dir:       r.FormValue("dir"),
-			Ref:       ref.String(),
-			Component: r.FormValue("component"),
+		if err := UnmarshalParams(r, p); err != nil {
+			util.WriteErrResponse(w, http.StatusBadRequest, err)
+			return
 		}
 
-		var res string
-		if err := h.Restore(p, &res); err != nil {
-			util.WriteErrResponse(w, http.StatusInternalServerError, err)
+		res, err := h.inst.Dispatch(r.Context(), method, p)
+		if err != nil {
+			util.RespondWithError(w, err)
 			return
 		}
-
 		util.WriteResponse(w, res)
+		return
+	}
+}
+
+// If the route has a dataset reference in the url, parse that ref, and
+// add it to the request object using the field "refstr".
+func addDsRefFromURL(r *http.Request, routePrefix string) error {
```

Comment:
Am not a huge fan of this, but this should go away once UnmarshalFromRequest is utilized across the board. Maybe just leave a TODO that also says so.

Reply:
Curious what you mean; which part are you not a fan of? The need for this function `addDsRefFromURL`? Or keeping the dsref in the url like `/status/dustmop/my_dataset`? From my understanding, UnmarshalFromRequest is basically a hook that lets parameters customize their conversion from http to concrete struct types. There's a lot of methods that allow dsrefs to live in the url, so I would prefer to avoid needing to repeat the same conversion code over and over if we can. I left a TODO comment somewhat to what effect on line 52 (the first usage of `addDsRefFromURL`), and plan on making a follow-up PR after this to demonstrate what I'm thinking.

Reply:
Oh, my bad, should have been more clear on this. 
So the reason I don't like it is that this further solidifies the pattern of prefix + hand parsing the ref at the end.
The intent of the comment was to steer this towards the recently added form of explicitly naming the dsref parts and using a common, global middleware to be super condensed on how we parse these.

In `api/api.go` we have

```
func handleRefRoute(m *mux.Router, ae lib.APIEndpoint, f http.HandlerFunc) {
	m.Handle(ae.String(), f)
	m.Handle(fmt.Sprintf("%s/%s", ae, "{peername}/{name}"), f)
	m.Handle(fmt.Sprintf("%s/%s", ae, "{peername}/{name}/{selector}"), f)
	m.Handle(fmt.Sprintf("%s/%s", ae, "{peername}/{name}/at/{fs}/{hash}"), f)
	m.Handle(fmt.Sprintf("%s/%s", ae, "{peername}/{name}/at/{fs}/{hash}/{selector}"), f)
}
```
And when we add add a route based ref handler we use the above form to define the route/handler mapping.
eg `handleRefRoute(m, lib.AEGet, s.Middleware(dsh.GetHandler))`

In `api/middleware` we then have
```
func refStringMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		setRefStringFromMuxVars(r)
		next.ServeHTTP(w, r)
	})
}
```
Which acts as the propagation engine for the refstr query param.
It will also make it easier to switch later on to attach to context/scope instead of query param.

Finally, the `addDsRefFromURL` fails on any url that has content/path components *after* the ref itself in this case. At the same time I was aiming to remove entirely `lib/api.go DsRefFromPath` and similar functions once the refactor is complete.

But to distill my own fluff, I would prefer we use the `handleRefRoute` approach.